File: bindings/node/samples/pubsub_recv_sample.ts
// SPDX-License-Identifier: MPL-2.0

'use strict';

const assert = require('node:assert/strict');
const { once } = require('node:events');
const net = require('node:net');
const zlink = require('../dist');

async function reservePort() {
  const srv = net.createServer();
  srv.listen(0, '127.0.0.1');
  await once(srv, 'listening');
  const { port } = srv.address();
  await new Promise((resolve, reject) => srv.close((error) => error ? reject(error) : resolve()));
  return port;
}

async function main() {
  const port = await reservePort();
  const endpoint = `tcp://127.0.0.1:${port}`;
  const ctx = new zlink.Context();
  const pub = new zlink.PubSocket(ctx);
  const sub = new zlink.SubSocket(ctx);

  try {
    const pubMonitor = pub.monitorOpen(zlink.MonitorEvent.CONNECTION_READY_CHANGED);
    const subMonitor = sub.monitorOpen(zlink.MonitorEvent.CONNECTION_READY_CHANGED);
    try {
      pub.bind(endpoint);
      sub.connect(endpoint);
      pubMonitor.recv();
      subMonitor.recv();
    } finally {
      pubMonitor.close();
      subMonitor.close();
    }

    const topic = 'prices';
    const sent = '101.25';
    sub.setSubscription(topic);
    pub.publish(topic, zlink.Message.copyOf(sent));

    const received = sub.subscribe();
    const recv = received.parts[0].toBuffer().toString();
    assert.equal(received.topic, topic);
    assert.equal(recv, sent);
    console.log(`[pubsub/recv] publish: "${topic}/${sent}" \u2192 subscribe: "${topic}/${recv}"`);
  } finally {
    sub.close();
    pub.close();
    ctx.close();
  }
}

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});

File: bindings/cpp/samples/pubsub_recv_sample.cpp
/* SPDX-License-Identifier: MPL-2.0 */

#include "sample_common.hpp"

int main ()
{
    zlink::context_t ctx;
    zlink::xpub_socket_t publisher (ctx);
    zlink::sub_socket_t subscriber (ctx);
    zlink::monitor_handle_t pub_monitor = publisher.monitor_handle ();
    zlink::monitor_handle_t sub_monitor = subscriber.monitor_handle ();

    assert (publisher.bind ("tcp://127.0.0.1:0") == 0);
    const std::string endpoint = publisher.options ().last_endpoint ();
    assert (!endpoint.empty ());
    assert (subscriber.connect (endpoint) == 0);
    assert (detail::wait_connected (pub_monitor, sub_monitor));

    const std::string topic = detail::k_pubsub_topic;
    subscriber.set_subscription (topic);

    const zlink::subscription_event_t event =
      publisher.receive_subscription_event ();
    assert (event.subscribed);
    assert (event.topic == topic);

    const std::string sent = detail::k_pubsub_payload;
    zlink::message_t outbound = detail::make_message (sent);
    publisher.publish (topic, outbound);

    const zlink::topic_message_t inbound = subscriber.subscribe ();
    assert (inbound.topic () == topic);
    assert (inbound.parts ().size () == 1);
    const std::string received = inbound.parts ()[0].to_string ();
    assert (received == detail::k_pubsub_payload);
    std::printf ("[pubsub/recv] publish: \"%s/%s\" → subscribe: \"%s/%s\"\n",
                 topic.c_str (), sent.c_str (), topic.c_str (), received.c_str ());
    return 0;
}
